File: Algo/GameOfLife.java
/* https://leetcode.com/problems/game-of-life/ */

class GameOfLife
{
	public static void main(String args[])
	{
		int[][] board = {
							{1, 1, 1},
							{1, 0, 0},
							{0, 0, 0}
						};
		GameOfLife g = new GameOfLife();
		g.gameOfLife(board);
		for(int[] row: board)
		{
			for(int col: row)
				System.out.print(col);
			System.out.println();
		}

	}

	public void gameOfLife(int[][] board) 
	{
		 if(board.length!=0 && board[board.length-1].length!=0)
        	getNextState(board, 0, 0);
    }

    public void getNextState(int [][] board, int i, int j)
    {
    	int row = board.length;
    	int col = board[board.length-1].length;

    	int countLiveNeighbors = 0;
    	int countDeadNeighbors = 0;
    	int nextState = -1;

    	if(i>row-1)
    		return;
    	if(i+1<row )
    	{
    		if(board[i+1][j]==1)countLiveNeighbors++;
    	}
    	if(i-1>-1)
    	{
    		if(board[i-1][j]==1)countLiveNeighbors++;
    	}
    	if(j+1<col)
    	{
    		if(board[i][j+1]==1)countLiveNeighbors++;
    	}
    	if(j-1>-1)
    	{
    		if(board[i][j-1]==1)countLiveNeighbors++;
    	}
    	if(i-1>-1 && j-1 >-1)
    	{
    		if(board[i-1][j-1]==1)countLiveNeighbors++;
    	}
    	if(i+1<row && j-1 >-1)
    	{
    		if(board[i+1][j-1]==1)countLiveNeighbors++;
    	}
    	if(i-1>-1 && j+1 < col)
    	{
    		if(board[i-1][j+1]==1)countLiveNeighbors++;
    	}
    	if(i+1<row && j+1 < col)
    	{
    		if(board[i+1][j+1]==1)countLiveNeighbors++;
    	}
    	if(board[i][j]==1)
    	{
    		if(countLiveNeighbors<2)
    			nextState = 0;
    		else if(countLiveNeighbors >= 2 && countLiveNeighbors <= 3)
    			nextState = 1;
    		else if(countLiveNeighbors > 3) 
    			nextState = 0;
    	}
    	else
    	{
    		if(countLiveNeighbors==3)
    			nextState = 1;
    		else
    			nextState = 0;
    	}
    	// System.out.println(nextState);
    	if(j+1==col)
    		getNextState(board, i+1, 0);
    	else
    		getNextState(board, i, j+1);

    	board[i][j]=nextState;
    }
}

File: Algo/python/game-of-life.py
''' https://leetcode.com/problems/game-of-life/ '''

class Solution(object):
    def getNextState(self, board, i, j, m, n):
        live_n = 0
        if (i-1) >= 0:
            if (j-1) >= 0:
                live_n += board[i-1][j-1]
            live_n += board[i-1][j]
            if (j+1) < n:
                live_n += board[i-1][j+1]
        if (i+1) < m:
            if (j-1) >= 0:
                live_n += board[i+1][j-1]
            live_n += board[i+1][j]
            if (j+1) < n:
                live_n += board[i+1][j+1]
        if (j-1) >= 0:
            live_n += board[i][j-1]
            
        if (j+1) < n:
            live_n += board[i][j+1]
            
        res = board[i][j]
        if res == 0 and live_n == 3:
            res = 1
        elif res == 1:
            if (live_n < 2) or (live_n > 3):
                res = 0
        #print board[i][j], i, j, res, live_n
        return res
        
    def gameOfLife(self, board):
        """
        :type board: List[List[int]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        m = len(board)
        n = len(board[0])
        board2=[]
        for i in range(m):
            board2.append([])
            for j in range(n):
                board2[i].append(board[i][j])
        
        for i in range(m):
            for j in range(n):
                board[i][j] = self.getNextState(board2, i, j, m, n)
        
